theurgy.lic

```diff
@@ -2,185 +2,453 @@
   Documentation: https://elanthipedia.play.net/Lich_script_repository#theurgy
 =end
 
-custom_require.call(%w[common common-items common-money common-travel drinfomon spellmonitor equipmanager])
+custom_require.call(%w[common common-arcana common-items common-money common-travel drinfomon spellmonitor equipmanager])
 
-class Theurgy
+class TheurgyActions
   include DRC
   include DRCI
-  include DRCM
   include DRCT
+  include DRCM
+  include DRCA
+  
+  def initialize(settings, data)
+    @settings = settings
+    @hometown = @settings.hometown
+    @data = data
+
+    @theurgy_supply_container = @settings.theurgy_supply_container
+    @water_holder = @settings.water_holder
+    @immortal_aspect = @settings.immortal_aspect
+    @flint_lighter = @settings.flint_lighter
+    @safe_room = @settings.safe_room
+
+    @rituals = get_rituals
+    @communes = get_communes
+
+    annotate_item_presence(@rituals + @communes)
+    
+    @rituals.reject! do |r|
+      r[:items].any? { |i| !i[:on_hand] && !i[:shop] }
+    end
+    @communes.reject! do |c|
+      c[:items].any? { |i| !i[:on_hand] && !i[:shop] }
+    end
+  end
 
-  ######################## GET/PUT ########################
-
-  def put_in_container(item)
-    bput "put #{item} in my #{@theurgy_supply_container}", 'You put', 'What were you referring to'
+  def get_rituals
+    rituals = all_rituals
+    blacklist = @settings.theurgy_blacklist
+    rituals.reject! { |r| blacklist.include? r[:method].to_s }
+    whitelist = @settings.theurgy_whitelist
+    unless whitelist.empty?
+      rituals.select! { |r| whitelist.include? r[:method].to_s }
+    end
+    rituals
+  end
+  
+  def all_rituals
+    rituals = []
+    if @data['almsbox'] && @settings.tithe
+      rituals << {
+        method: :tithe,
+      }
+    end
+    if @hometown == 'Crossing'
+      if DRStats.circle >= 30
+        rituals << {
+          method: :study_wall,
+        }
+      end
+      rituals << {
+        method: :refectory,
+      }
+    end
+    if DRSkill.getrank('Mechanical Lore') > 140 && @immortal_aspect && @data['limb_foraging']
+      # TODO: allow other knife nouns 
+      rituals << {
+        method: :carve_bead,
+        items: [
+          { name: 'carving knife', any_container: true },
+          { name: 'shaper', any_container: true},
+          { name: 'prayer chain', any_container: true },
+        ],
+      }
+    end
+    if @data['gather_sirese'] && @data['plant_sirese']
+      rituals << {
+        method: :sirese_seed,
+      }
+    end
+    if @data['bath']
```

Comment:
This is a long process and entirely not needed. I vote that bathing be removed from Theurgy altogether.

Reply:
Also, was support for this checked in Shard? I remember having an issue finding an altar which supported all of the same things that the Altar in the Crossing supports.

Reply:
Not supported in Dhard.  This only runs if there's `@data['bath']` set for the region.  Note that this can be blacklisted if not wanted -- for my lowbie cleric it was helpful for hitting mindlock since I don't have all the communes.

Reply:
Yeah this was a really useful ritual starting out, even though its not useful at 150. Maybe blacklisting it by default above a set number of theurgy ranks or something would work? That would be something for another pull request though.

Reply:
Gotcha. I wonder if a warning or comment would be good for these in the YAMLs. It's not blatantly obvious which are and aren't supported in what towns based on YAML alone, you'd need to be able to read the code.

Reply:
I stopped using bathe WAY before 150. It's not useful pretty early on, it takes very few communes to cap Theurgy after they fixed it a few months back.

Reply:
Very early this was helpful to me (i.e. before I had > 2 communes).  It's still a good devotion regen.  I'd like to leave it in and allow blacklisting.

I do agree it's not obvious from the YAML what's available.  I will add all the existing rituals to the whitelist with comments describing them.